### DISCUSSION TITLE
Can zip.js support safari browser unZip?

### DISCUSSION THREAD

**Discussion by BarryJame** [gildas-lormeau/zip.js#247] [Q&A] [answered]
When I Quote u Zip.js ,I found it can work very well except iOS Safari Browser,it shows below.
![image](https://user-images.githubusercontent.com/25240047/117923503-b747d780-b326-11eb-963d-c44621f5584f.png)

**Comment by gildas-lormeau** [comment, marked as answer]
As I write this, Safari does not support [`DataView#getBigUint64`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/DataView/getBigUint64) (see [here](https://caniuse.com/?search=getBigUint64)) so it must be polyfilled if you want to read files using zip64. Here is below an implementation based on [this response from Stack Overflow](https://stackoverflow.com/questions/53103695/how-to-read-64-bit-integer-from-an-arraybuffer-dataview-in-javascript#answer-53107482). You have to run this script before using zip.js.

```js
if (DataView.prototype.getBigUint64 === undefined) {
  DataView.prototype.getBigUint64 = function(byteOffset, littleEndian) {
    // split 64-bit number into two 32-bit parts
    const left =  this.getUint32(byteOffset, littleEndian);
    const right = this.getUint32(byteOffset+4, littleEndian);

    // combine the two 32-bit values
    const combined = littleEndian? left + 2**32*right : 2**32*left + right;

    if (!Number.isSafeInteger(combined)) {
      console.warn(combined, 'exceeds MAX_SAFE_INTEGER. Precision may be lost');
    }

    return combined;
  };
}
```

**Reply by BarryJame** [reply]
thks boss

**Comment by phamconganh** [comment]
@gildas-lormeau, i get the same issue but it is view.getUnit16 on safari. Can you tell me how to fix this?
`TypeError: undefined is not an object (evaluating 't.getUint16')`

**Reply by gildas-lormeau** [reply]
This is not the same error (`'t.getBigUint64' is not a function` vs `undefined is not an object`). I need at least a complete stack trace to help you. Please also provide the version of Safari.

**Reply by phamconganh** [reply]
sorry. I double check my build and my problem was with the build options.
I fixed the build options then it ran.
And I have a question. zipjs slow performance in safari 12.2 OS?
My ipad mini is OS 12.2 and zip.js is not running.
I test on safari iphoneX, then it is running.

**Reply by gildas-lormeau** [reply]
Is it just slow or not running at all?

**Reply by phamconganh** [reply]
I test on safari 12.2, it's not running. On firefox it's slow. On safari 15 and 16 it's not much different from normal, about 50ms slow

**Reply by phamconganh** [reply]
On firefox slow about 500ms, on chrome 70+ it's pretty fast, about 30ms. All zip test data from 200Kb with password

**Reply by gildas-lormeau** [reply]
Do you see any errors when running on Safari? For example zip.js relies on the `TransformStream` API which is available on Safari since the version 14.1 (see https://developer.mozilla.org/en-US/docs/Web/API/TransformStream/TransformStream#browser_compatibility). If you or your build don't provide a polyfill for this API, it won't work.

**Reply by gildas-lormeau** [reply]
In addition, I guess the slowness comes from the fact that zip.js uses the `DecompressionStream` native API which is not available (yet?) on Firefox.
